File: qqbot-module-spi/src/main/java/com/mieai/qqbot/module/spi/FrameworkModule.java
package com.mieai.qqbot.module.spi;

import com.mieai.qqbot.module.api.ModuleDescriptor;
import java.util.Objects;

/** Optional lifecycle entry point supplied by a startup-loaded framework module JAR. */
public interface FrameworkModule {
    ModuleDescriptor descriptor();

    default void start(ModuleContext context) {}

    default void stop() {}

    static FrameworkModule declarative(ModuleDescriptor descriptor) {
        ModuleDescriptor value = Objects.requireNonNull(descriptor, "descriptor must not be null");
        return () -> value;
    }
}

File: qqbot-module-spi/src/main/kotlin/com/mieai/qqbot/module/spi/FrameworkModule.kt
package com.mieai.qqbot.module.spi

import com.mieai.qqbot.module.api.ModuleDescriptor

/** Optional lifecycle entry point supplied by a startup-loaded framework module JAR. */
interface FrameworkModule {
    val descriptor: ModuleDescriptor

    fun start(context: ModuleContext) {}

    fun stop() {}

    companion object {
        fun declarative(value: ModuleDescriptor): FrameworkModule = object : FrameworkModule {
            override val descriptor = value
        }
    }
}
